backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java

```diff
@@ -0,0 +1,267 @@
+import de.metas.i18n.AdMessageKey;
+import de.metas.i18n.IMsgBL;
+import de.metas.lang.SOTrx;
+import de.metas.order.IOrderBL;
+import de.metas.order.OrderId;
+import de.metas.organization.IOrgDAO;
+import de.metas.organization.LocalDateAndOrgId;
+import de.metas.organization.OrgId;
+import de.metas.product.ProductId;
+import de.metas.quantity.Quantity;
+import de.metas.uom.IUOMDAO;
+import de.metas.uom.UomId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseBL;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_C_OrderLine;
+import org.compiere.model.I_C_UOM;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Component
+public class SalesOrderLineModularContractHandler implements IModularContractTypeHandler<I_C_OrderLine>
+{
+	private static final AdMessageKey MSG_ON_COMPLETE_DESCRIPTION = AdMessageKey.of("de.metas.contracts.modular.impl.SalesOrderLineModularContractHandler.OnComplete.Description");
+	private static final AdMessageKey MSG_ON_REVERSE_DESCRIPTION = AdMessageKey.of("de.metas.contracts.modular.impl.SalesOrderLineModularContractHandler.OnReverse.Description");
+	private static final AdMessageKey MSG_ERR_DELETION_NOT_ALLOWED = AdMessageKey.of("de.metas.contracts.modular.impl.SalesOrderLineModularContractHandler.DeletionNotAllowed");
+
+	private final IMsgBL msgBL = Services.get(IMsgBL.class);
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final IUOMDAO uomDAO = Services.get(IUOMDAO.class);
+	private final IOrderBL orderBL = Services.get(IOrderBL.class);
+	private final IWarehouseBL warehouseBL = Services.get(IWarehouseBL.class);
+	private final IFlatrateBL flatrateBL = Services.get(IFlatrateBL.class);
+
+	private final ModularContractSettingsDAO modularContractSettingsDAO;
+	private final ModularContractLogDAO contractLogDAO;
+
+	public SalesOrderLineModularContractHandler(
+			@NonNull final ModularContractSettingsDAO modularContractSettingsDAO,
+			@NonNull final ModularContractLogDAO contractLogDAO)
+	{
+		this.modularContractSettingsDAO = modularContractSettingsDAO;
+		this.contractLogDAO = contractLogDAO;
+	}
+
+	@Override
+	public @NonNull Class<I_C_OrderLine> getType()
+	{
+		return I_C_OrderLine.class;
+	}
+
+	@Override
+	public boolean applies(final @NonNull I_C_OrderLine orderLine)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+		if (SOTrx.ofBoolean(order.isSOTrx()).isPurchase())
+		{
+			return false;
+		}
+
+		final BPartnerLocationAndCaptureId orderBillPartnerId = orderBL.getBillToLocationId(order);
+
+		final boolean isMatchingPartner = Optional.ofNullable(WarehouseId.ofRepoIdOrNull(order.getM_Warehouse_ID()))
```

Comment:
I don't think we need this validation. We can sell to any customer as long as the M_Warehouse.C_BPartner_ID is the one with the contract

Reply:
not needed actually. will fix it